src/app/recipes/components/recipe-details/recipe-details.component.ts

```diff
@@ -54,9 +64,113 @@ export class RecipeDetailsComponent implements OnInit, OnDestroy {
         } else {
             this.user.favoriteRecipes = [...this.user.favoriteRecipes, this.recipe.id];
         }
+
+        this.recipeBookmarked = this.isRecipeBookmarked();
+    }
+
+    isRecipeBookmarked() {
+       return this.user.favoriteRecipes.includes(this.recipe.id);
+    }
+
+    scroll(el: HTMLElement) {
+        el.scrollIntoView({behavior: 'smooth'});
+    }
+
+    onCheckIngredient(category: string, ingredient: string) {
+        let ingredients: CheckedRecipeIngredients[] = [];
+        if (this.isLocalStorageAvailable) {
+            ingredients = JSON.parse(localStorage.getItem(this.userCheckedIngredientsKey) || '[]');
+        }
+
+        let recipeIndex: number = ingredients.findIndex(item => item.recipeId === this.recipe.id);
+        const categories: IngredientCategory[] = ingredients[recipeIndex]?.categories;
+
+        if (ingredients.length === this.maxStoredRecipes && recipeIndex === -1) {
+            ingredients.splice(this.getOldestEditedRecipeIndex(ingredients), 1);
+        }
+
+        if (recipeIndex === -1 && ingredients.length < this.maxStoredRecipes) {
+            ingredients.push({
+                recipeId: this.recipe.id,
+                lastEditDate: new Date().toString(),
+                categories: [{
+                    name: category,
+                    items: [ingredient]
+                }]
+            });
+        } else {
+            const categoryIndex: number = categories.findIndex(cat => cat.name === category);
+            const selectedCategory: IngredientCategory = categories[categoryIndex];
+
+            if (categoryIndex === -1) {
+                ingredients[recipeIndex].categories.push({
+                    name: category,
+                    items: [ingredient]
+                });
+            } else {
+                const itemIndex: number = selectedCategory.items.findIndex(item => item === ingredient);
+
+                itemIndex === -1 ? selectedCategory.items.push(ingredient) : selectedCategory.items.splice(itemIndex, 1);
+
+                if (selectedCategory.items.length === 0) {
+                    categories.splice(categoryIndex, 1);
+                }
+            }
+
+            ingredients[recipeIndex].lastEditDate = new Date().toString();
+            if (categories.length === 0) {
+                ingredients.splice(recipeIndex, 1);
+            }
+        }
+
+        if (this.isLocalStorageAvailable) {
+            if (ingredients.length > 0) {
+                localStorage.setItem(this.userCheckedIngredientsKey, JSON.stringify(ingredients));
+            } else {
+                localStorage.removeItem(this.userCheckedIngredientsKey);
+            }
+        }
+        this.getFromStorage();
+        this.updateShowClearAllStatus();
```

Comment:
Maybe after extracting it to service this could be optimized/split. Now it looks complex although this storage task is kind of simple